### ISSUE TITLE
A quick question on DPTNet implementation

### ISSUE THREAD

**Issue by cliffzhao** [asteroid-team/asteroid#380]
Hi, 

I have a quick question about the implementation of improved transformer in DPTNet. I quote the part of current implementation as below. 

https://github.com/asteroid-team/asteroid/blob/3a458ca3aed2a11b8b26dbeaf0aceb0f57b4b04d/asteroid/masknn/attention.py#L57-L59

As the comment shows, x is [batch, seq_len, channels], and it is fed to self.mha (MultiheadAttention in PyTorch). However, according to the docs of MultiheadAttention, the input is [seq_len, batch, channels] in the default setup.

Is there any bug here? Thanks!

**Comment by mpariente** [comment]
I just checked, and I agree with you, this doesn't seem congruent. 
@popcornell could you have a look please?

**Comment by popcornell** [comment]
Definitely a bug, i did not catch it because the net was training strangely ( i did not have time/resources to run a full training unfortunately).
Thank you very much @cliffzhao, I made a PR which should fix this.

**Comment by cliffzhao** [comment]
> Definitely a bug, i did not catch it because the net was training strangely ( i did not have time/resources to run a full training unfortunately).
> Thank you very much @cliffzhao, I made a PR which should fix this.

Great! Thanks @popcornell and @mpariente. At the same time, we also need to be careful with RNN's input too.

**Comment by popcornell** [comment]
@cliffzhao I merged it I think now it should be fine

**Comment by mpariente** [comment]
Though it's weird, I think @b02901071 managed to train a full model.. 
See [this paper](https://arxiv.org/pdf/2010.15366.pdf)
I asked if they changed anything to fix it. 

Thanks for the feedback @cliffzhao

**Comment by b02901071** [comment]
> Definitely a bug, i did not catch it because the net was training strangely ( i did not have time/resources to run a full training unfortunately).
> Thank you very much @cliffzhao, I made a PR which should fix this.

I didn't mention it either, since it just trained without any error... but not sure whether it affects the performance or not. I trained with multi-gpu and accumulated grad, so the batch size on each gpu is just 1, maybe that's the reason why it could be trained without errors.

**Comment by mpariente** [comment]
But that would mean that the attention is over the batch_size and the sequence element are treated independently e.g no attention at all.. right?

**Comment by SungFeng-Huang** [comment]
Probably... and since the batch_size of the input on each gpu is 1, the mha might view the input as a tensor with length 1, so the mha actually does nothing and just pass the input forward.

**Comment by popcornell** [comment]
Since the model is quite heavy on memory I also did only train only for a few epochs with batch size 1. 
The MHA reduces to just learned domain cross-correlation of the representation at time-step X with itself and then it just rescores a transformed version of it by softmax. The self attention is computed independently for each timestep.
We have done an ablation study without knowing (sorry again !), if @SungFeng-Huang have you attained more or less same performance, well then the temporal modelling is just done by LSTMs and MHA (implemented properly) does not help (it helps only its diagonal actually as here seems). 
This is very interesting maybe we could even make a paper on these things.

**Comment by popcornell** [comment]
Am I hallucinating or even the original DPTNet also does this error ?
https://github.com/ujscjj/DPTNet/blob/master/transformer_improved.py

**Comment by SungFeng-Huang** [comment]
> Am I hallucinating or even the original DPTNet also does this error ?
> https://github.com/ujscjj/DPTNet/blob/master/transformer_improved.py

Original DPTNet seems to do the permutation before forwarding to the layer:
https://github.com/ujscjj/DPTNet/blob/987b6107012b21105f6ec0cdd8b5c2d3b250ae1d/models.py#L15

**Comment by popcornell** [comment]
Missed that bit, so also here https://github.com/asteroid-team/asteroid/blob/f5939edafc2811988881c8a9f8d53120b2efb615/asteroid/masknn/attention.py#L60
we miss the FeedForward network after MHA (which to me does not make very much sense since afterwards there is a LSTM). 
Thanks @SungFeng-Huang

**Comment by mpariente** [comment]
Where is the double dropout? 

Their code also seems weird to me. The LSTM is applied to the permuted input? Which means that the batch is treated as a sequence?? Am I missing something?

**Comment by popcornell** [comment]
They apply after the MHA a MLP for each timestep (standard transformer encoder). Then they apply LSTM. The MLP seems to me a waste of parameters a lil bit. Gonna add that however

**Comment by mpariente** [comment]
But the LSTM is applied to the output of `MultiheadAttention` which is `(sequence, batch, embed_dim)` right?

**Comment by mpariente** [comment]
Ok, at first this didn't seem right, but I forgot that by default `nn.LSTM` has `(seq_len, batch, input_size)` as input size so their implementation looks ok. 

However, ours doesn't because we permute back before the recurrent layer, and we shouldn't, right?

**Comment by popcornell** [comment]
Forget my last two comments, the LSTM was wrong also, which means that the whole architecture was a giant MLP basically

**Comment by popcornell** [comment]
@cliffzhao @SungFeng-Huang  can you check now this https://github.com/asteroid-team/asteroid/pull/383 ?

**Comment by SungFeng-Huang** [comment]
> @cliffzhao @SungFeng-Huang can you check now this #383 ?

Approved.

**Comment by cliffzhao** [comment]
> > Am I hallucinating or even the original DPTNet also does this error ?
> > https://github.com/ujscjj/DPTNet/blob/master/transformer_improved.py
> 
> Original DPTNet seems to permute before forwarding to the layer:
> https://github.com/ujscjj/DPTNet/blob/987b6107012b21105f6ec0cdd8b5c2d3b250ae1d/models.py#L15

Hi @SungFeng-Huang , I checked this before. In the released code segments from the author, I think he does the right thing. He employs the default setup of MHA and LSTM in PyTorch, which is [seq_len, batch_size, channels]. That's why he permute the input before feeding to his module "self.transformer".

> Since the model is quite heavy on memory I also did only train only for a few epochs with batch size 1.
> The MHA reduces to just learned domain cross-correlation of the representation at time-step X with itself and then it just rescores a transformed version of it by softmax. IE: the self attention is computed independently for each timestep, you get only the diagonal of self attention.
> We have done an ablation study without knowing (sorry again !), if @SungFeng-Huang have you attained more or less same performance, well then the temporal modelling is just done by LSTMs and MHA (implemented properly) does not help (it helps only its diagonal actually as here seems).
> This is very interesting maybe we could even make a paper on these things.

Interesting thing is that, if we set all the input shape to MHA and LSTM correctly, the performance is not good. I implemented DPTNet before in Asteroid by myself and incautiously swapped seq_len and batch_size in LSTM, I got reasonable good performance. But when I fix it in my code and keep everything correctly, it becomes much worse. That's why I back to the DPTNet implementation in Asteroid and find it strange in the input shape. In my experiments, the batch size is 2 in multi-GPU training.

@SungFeng-Huang if you have time, could you please do a quick check here? Keep everything same as before, just simply fix the input shape issue, and run a few epochs to make a comparison.

**Comment by popcornell** [comment]
> Interesting thing is that, if we set all the input shape to MHA and LSTM correctly, the performance is not good. I implemented DPTNet before in Asteroid by myself and incautiously swapped seq_len and batch_size in LSTM, I got reasonable good performance. But when I fix it in my code and keep everything correctly, it becomes much worse. That's why I back to the DPTNet implementation in Asteroid and find it strange in the input shape. In my experiments, the batch size is 2 in multi-GPU training.

Oh wow this is super interesting and also super strange !! I also did not notice the error at all because the performance was pretty decent even if, in principle, it should not work. I mean both MHA and LSTMs were wrong and unable to do temporal modelling. Pretty much all literature in neural source sep suggests that temporal modelling is key to get good performance and in fact it is the whole reason behind DPTNet and DPRNN. I am puzzled a bit.

**Comment by cliffzhao** [comment]
> > Interesting thing is that, if we set all the input shape to MHA and LSTM correctly, the performance is not good. I implemented DPTNet before in Asteroid by myself and incautiously swapped seq_len and batch_size in LSTM, I got reasonable good performance. But when I fix it in my code and keep everything correctly, it becomes much worse. That's why I back to the DPTNet implementation in Asteroid and find it strange in the input shape. In my experiments, the batch size is 2 in multi-GPU training.
> 
> Oh wow this is super interesting and also super strange !! I also did not notice the error at all because the performance was pretty decent even if, in principle, it should not work. I mean both MHA and LSTMs were wrong and unable to do temporal modelling. Pretty much all literature in neural source sep suggests that temporal modelling is key to get good performance and in fact it is the whole reason behind DPTNet and DPRNN. I am puzzled a bit.

Yeah, me too. We have correct implementation of DPRNN in Asteroid (no input shape issue) and I also simply tried Facebook's gated DPRNN based on Asteroid DPRNN's implementation. Both of them work well as expected. But when it comes to DPTNet, some strange thing happens.

Let's see @SungFeng-Huang 's  quick check when his time is available. If similar phenomenon can be observed (performance drops significantly after fixing the input shape issue), that would be quite interesting and it is worth an in-depth study.

**Comment by popcornell** [comment]
If you have Facebook Gated DPRNN, when you have time it will be nice if you can do a pull-request to add it. 
It would be cool to have in the toolkit !

**Comment by cliffzhao** [comment]
> If you have Facebook Gated DPRNN, when you have time it will be nice if you can do a pull-request to add it.
> It would be cool to have in the toolkit !

Sure. Currently I didn't do a full implementation of their paper. I just simply tried their proposed gated DPRNN structure, there is no speaker embedding loss and multi-layer loss. If I get them done, I'm glad to do a pull-request. I did learn a lot from Asteroid in recent usage, and I will be happy to contribute. Thanks all the developers to build Asteroid.

**Comment by SungFeng-Huang** [comment]
I've trained 5 epochs with the correct version, and the val_loss is just slightly better than the original one:

epoch|1|2|3|4|5
---|---|---|---|---|---
original|-3.18|-5.23|-6.68|-7.83|-8.92
correct|-3.81|-5.64|-7.09|-8.43|-9.42

Probably means that the MHA is not necessary?

**Comment by cliffzhao** [comment]
> I've trained 5 epochs with the correct version, and the val_loss is just slightly better than the original one:
> 
> epoch	1	2	3	4	5
> original	-3.18	-5.23	-6.68	-7.83	-8.92
> correct	-3.81	-5.64	-7.09	-8.43	-9.42
> Probably means that the MHA is not necessary?

Thanks @SungFeng-Huang for your quick check!

If we remove the MHA, is it almost the same as DPRNN? More interesting question is, why the previous incorrect one works also well? I plan to do some additional experiments at my end, especially with 1 batch size. I will be back to this issue after I get some results.

**Comment by mpariente** [comment]
FYI, the facebook paper is implemented [here](https://github.com/facebookresearch/svoice), but it's under ` CC-BY-NC-SA 4.0` so we cannot copy paste anything. 
Maybe we could be inspired from it though.

**Comment by SungFeng-Huang** [comment]
> > I've trained 5 epochs with the correct version, and the val_loss is just slightly better than the original one:
> > epoch	1	2	3	4	5
> > original	-3.18	-5.23	-6.68	-7.83	-8.92
> > correct	-3.81	-5.64	-7.09	-8.43	-9.42
> > Probably means that the MHA is not necessary?
> 
> Thanks @SungFeng-Huang for your quick check!
> 
> If we remove the MHA, is it almost the same as DPRNN? More interesting question is, why the previous incorrect one works also well? I plan to do some additional experiments at my end, especially with 1 batch size. I will be back to this issue after I get some results.

I agree that removing MHA should be almost the same as DPRNN. But this reminds me of the weird cuda memory usage of the incorrect DPTNet, which is much more than DPRNN.
Since the model sizes of DPTNet and DPRNN are close, I believe the memory usage mainly depends on computing the attention map of the MHA, and unrolling the LSTM. However, the incorrect DPTNet only generates 1x1 attention map for each time step, and the LSTM is fed with input size (1, seq_len, channel) without setting batch_first to True, so the LSTM doesn't do any recurrent stuff at all. Under these situations, I cannot imagine why the incorrect DPTNet uses that much cuda memory?

**Comment by cliffzhao** [comment]
> > > I've trained 5 epochs with the correct version, and the val_loss is just slightly better than the original one:
> > > epoch	1	2	3	4	5
> > > original	-3.18	-5.23	-6.68	-7.83	-8.92
> > > correct	-3.81	-5.64	-7.09	-8.43	-9.42
> > > Probably means that the MHA is not necessary?
> > 
> > 
> > Thanks @SungFeng-Huang for your quick check!
> > If we remove the MHA, is it almost the same as DPRNN? More interesting question is, why the previous incorrect one works also well? I plan to do some additional experiments at my end, especially with 1 batch size. I will be back to this issue after I get some results.
> 
> I agree that removing MHA should be almost the same as DPRNN. But this reminds me of the weird cuda memory usage of the incorrect DPTNet, which is much more than DPRNN.
> Since the model sizes of DPTNet and DPRNN are close, I believe the memory usage mainly depends on computing the attention map of the MHA, and unrolling the LSTM. However, the incorrect DPTNet only generates 1x1 attention map for each time step, and the LSTM is fed with input size (1, seq_len, channel) without setting batch_first to True, so the LSTM doesn't do any recurrent stuff at all. Under these situations, I cannot imagine why the incorrect DPTNet uses that much cuda memory?

Hi @SungFeng-Huang , I think the problem is coming from intra-processing and inter-processing. For each processing block, it takes [batch_size, feats, chunk_size, num_chunks] in, let's use [B, N, K, L] for simplicity.

For intra-processing, we are dealing with [B * L, K, N]; for inter-processing, it is [B * K, L, N]. So even if the batch_size (B) is set to 1, the actual number that MHA/LSTM deals with is not 1. So it won't be 1x1 if my understanding is correct.

**Comment by SungFeng-Huang** [comment]
Oh, I see! Thanks @cliffzhao for the explanation. And I think that could be the reason why the incorrect DPTNet still works as well. For intra-processing, we should input [K, 1*L, N] to the MHA but actually we input [1*L, K, N] instead, which is the input of inter-processing, and vice versa. This means that the incorrect DPTNet is actually a DPTNet with the opposite order of inter- and intra-processing.

**Comment by mpariente** [comment]
Oh, yes that makes sense!
And the same applies to the LSTM, so it just switched intra- and inter-processing, that shouldn't have too much effect. Nice detective work everyone :wink:

**Comment by popcornell** [comment]
Thank you @cliffzhao @SungFeng-Huang  mystery solved then !
